BrowserKit/Sources/WebEngine/Utilities/URLFormatter.swift

```diff
@@ -73,17 +73,46 @@ public final class DefaultURLFormatter: URLFormatter {
             }
         }
 
-        guard let url = URL(string: entry) else { return nil }
-
         // Only allow this URL if it's safe
         let browsingContext = BrowsingContext(type: .internalNavigation, url: url)
         if securityManager.canNavigateWith(browsingContext: browsingContext) == .allowed {
-            return URL(string: entry)
+            return url
         } else {
             return nil
         }
     }
 
+    private func handleURL(with entry: String) -> URL? {
+        // Per Apple docs, for apps linked on or after iOS 17 and aligned OS versions, URL parsing has updated
+        // from the obsolete RFC 1738/1808 parsing to the same RFC 3986 parsing as URLComponents.
+        // iOS 17+, URL automatically percent- and IDNA-encodes invalid characters to help create a valid URL.
+
+        if #available(iOS 17, *) {
+            return URL(string: entry)
+        } else {
+            // We only want to do percent encoding on our other schemes
+            let hasHTTPScheme = entry.hasPrefix("http://") || entry.hasPrefix("https://")
+            guard !hasHTTPScheme else { return URL(string: entry) }
+
+            // FXIOS-14375 - Manually add percent encoding since it's not handled by URL(string:) in < iOS 17
+            guard let escapedURL = entry.addingPercentEncoding(withAllowedCharacters: urlAllowed) else {
+                return nil
+            }
+            guard let finalURL = URL(string: escapedURL), schemeIsValid(for: finalURL) else { return nil }
+            return finalURL
+        }
+    }
+
+    /**
+     Returns whether the URL's scheme is one of those listed on the official list of URI schemes.
+     This only accepts permanent schemes: historical and provisional schemes are not accepted.
+     */
+    public func schemeIsValid(for url: URL) -> Bool {
```

Comment:
Doesn't need to be public?
```suggestion
    func schemeIsValid(for url: URL) -> Bool {
```